sharding-core/sharding-core-preprocessor/src/test/java/org/apache/shardingsphere/core/optimize/segment/insert/InsertValueContextTest.java

```diff
@@ -0,0 +1,163 @@
+        ParameterMarkerExpressionSegment parameterMarkerExpressionSegment = new ParameterMarkerExpressionSegment(0, 10, 5);
+
+        return Lists.newArrayList(parameterMarkerExpressionSegment);
+    }
+
+    @Test
+    public void assertGetValueWhenLiteralExpressionSegment() {
+        Object literalObject = new Object();
+        Collection<? extends ExpressionSegment> assignments = makeLiteralExpressionSegment(literalObject);
+        List<? extends Object> parameters = Lists.newArrayList();
+
+        InsertValueContext insertValueContext = new InsertValueContext(assignments,parameters, 0);
+
+        Object result = insertValueContext.getValue(0);
+
+        assertThat(result, is(literalObject));
+    }
+
+    private Collection<LiteralExpressionSegment> makeLiteralExpressionSegment(Object literalObject) {
+        LiteralExpressionSegment parameterMarkerExpressionSegment = new LiteralExpressionSegment(0, 10, literalObject);
+
+        return Lists.newArrayList(parameterMarkerExpressionSegment);
+    }
+
+    @Test
+    public void assertAppendValueWhenParametersIsEmpty() {
+        Collection<? extends ExpressionSegment> assignments = Lists.newArrayList();
+        List<? extends Object> parameters = Lists.newArrayList();
+
+        InsertValueContext insertValueContext = new InsertValueContext(assignments,parameters, 0);
+
+        Object value = "test";
+        String type = "String";
+
+        insertValueContext.appendValue(value, type);
+
+        List<ExpressionSegment> valueExpressions = insertValueContext.getValueExpressions();
+        assertThat(valueExpressions.size(), is(1));
+
+        DerivedLiteralExpressionSegment segmentInInsertValueContext = (DerivedLiteralExpressionSegment) valueExpressions.get(0);
+        assertThat(segmentInInsertValueContext, is(new DerivedLiteralExpressionSegment(value, type)));
+    }
+
+    @Test
+    public void assertAppendValueWhenParametersIsNotEmpty() {
+        Collection<? extends ExpressionSegment> assignments = makeParameterMarkerExpresstionSegment();
+        String parameterValue = "test";
+        List<? extends Object> parameters = Lists.newArrayList(parameterValue);
+        int parametersOffset = 0;
+
+        InsertValueContext insertValueContext = new InsertValueContext(assignments,parameters, parametersOffset);
+
+        Object value = "test";
+        String type = "String";
+
+        insertValueContext.appendValue(value, type);
+
+        List<ExpressionSegment> valueExpressions = insertValueContext.getValueExpressions();
+        assertThat(valueExpressions.size(), is(2));
+
+        DerivedParameterMarkerExpressionSegment segmentInInsertValueContext = (DerivedParameterMarkerExpressionSegment) valueExpressions.get(1);
+        assertThat(segmentInInsertValueContext, is(new DerivedParameterMarkerExpressionSegment(parameters.size() - 1, type)));
+    }
+}
+
+
+@RequiredArgsConstructor
+class MethodInvocation<T> {
```

Comment:
What is the proposal for inner class?

Reply:
just for test convenience

Reply:
We may not add a class only for convenience, may need more design or inline into original class